payments/views.py

```diff
@@ -38,31 +38,63 @@ def get(self, request, *args, **kwargs):
 
 
 class WorkshopList(ListView):
-    queryset = Workshop.objects.filter(draft=False)
     context_object_name = 'upcoming_workshops'
+    private_code = None
+
+    # Since there are possible GET parameters we need to conditionally
+    # determine the queryset for the ListView
+    def get_queryset(self):
+        queryset = None
+        if self.private_code:
+            queryset = Workshop.objects.filter(private_code=self.private_code)
+
+        if queryset is None or len(queryset) == 0:
+            queryset = Workshop.objects.filter(draft=False, public=True)
+        return queryset
+
+    # Capture the GET and store it in the user session. This allows
+    # for blocking the ability to see a workshop detail page even if you know
+    # the slug for it.
+    def get(self, request, *args, **kwargs):
+        self.private_code = request.GET.get('code')
+        if self.private_code:
+            request.session['private_code'] = self.private_code
+        return super().get(request, *args, **kwargs)
 
 
 class WorkshopDetail(FormMixin, DetailView):
     model = Workshop
     form_class = OrderForm
     context_object_name = 'workshop'
+    discount_code = None
+
+    # dicount_code must be grabbed before anything begins as putting it
+    # in the get() was somehow being beaten by the get_form_kwargs()
+    def dispatch(self, request, *args, **kwargs):
+        self.discount_code = request.GET.get('rate')
+        return super().dispatch(request, *args, **kwargs)
 
     def get(self, request, *args, **kwargs):
         response = super().get(request, *args, **kwargs)
-        if not request.user.is_authenticated() and self.object.draft:
+        code = request.session.get('private_code')
+
+        if (not request.user.is_authenticated() and self.object.draft) or \
+           (not self.object.public and code != self.object.private_code):
             return HttpResponseRedirect(reverse('payments:index'))
         return response
 
     def get_form_kwargs(self):
         kwargs = super().get_form_kwargs()
         kwargs['workshop'] = self.object
+        kwargs['discount_code'] = self.discount_code
         return kwargs
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
```

Comment:
Before calling `super()` just do `self.discount_code = self.request.GET.get('rate')`, then you should have the discount code everywhere you need it.